kafka-rest/src/main/java/io/confluent/kafkarest/entities/v3/ProduceRequest.java

```diff
@@ -0,0 +1,305 @@
+    public abstract JsonNode getData();
+
+    public static Builder builder() {
+      return new AutoValue_ProduceRequest_ProduceRequestData.Builder();
+    }
+
+    @JsonCreator
+    static ProduceRequestData fromJson(
+        @JsonProperty("schema_subject_strategy") @Nullable
+            SchemaSubjectStrategy schemaSubjectStrategy,
+        @JsonProperty("type") @Nullable EmbeddedFormat format,
+        @JsonProperty("schema_subject") @Nullable String schemaSubject,
+        @JsonProperty("schema_id") @Nullable Integer schemaId,
+        @JsonProperty("schema_version") @Nullable Integer schemaVersion,
+        @JsonProperty("schema") @Nullable String rawSchema,
+        @JsonProperty("data") JsonNode data) {
+      return builder()
+          .setFormat(format)
+          .setSchemaSubjectStrategy(schemaSubjectStrategy)
+          .setSchemaSubject(schemaSubject)
+          .setSchemaId(schemaId)
+          .setSchemaVersion(schemaVersion)
+          .setRawSchema(rawSchema)
+          .setData(data)
+          .build();
+    }
+
+    @AutoValue.Builder
+    public abstract static class Builder {
+
+      Builder() {
+      }
+
+      public abstract Builder setFormat(@Nullable EmbeddedFormat format);
+
+      public abstract Builder setSchemaSubjectStrategy(
+          @Nullable SchemaSubjectStrategy schemaSubjectStrategy);
+
+      public abstract Builder setSchemaSubject(@Nullable String schemaSubject);
+
+      public abstract Builder setSchemaId(@Nullable Integer getSchemaId);
+
+      public abstract Builder setSchemaVersion(@Nullable Integer schemaVersion);
+
+      public abstract Builder setRawSchema(@Nullable String rawSchema);
+
+      public abstract Builder setData(JsonNode data);
+
+      abstract ProduceRequestData autoBuild();
+
+      public final ProduceRequestData build() {
+        ProduceRequestData request = autoBuild();
+
+        checkState(
+            !request.getSchemaSubjectStrategy().isPresent()
+                || !request.getSchemaSubject().isPresent(),
+            "Only one of 'schema_subject_strategy' or 'schema_subject' can be used.");
+
+        checkState(
+            request.getSchemaId().isPresent()
+                ? (!request.getSchemaVersion().isPresent() && !request.getRawSchema().isPresent())
+                : (!request.getSchemaVersion().isPresent() || !request.getRawSchema().isPresent()),
+            "Only one of 'schema_id', 'schema_version' or 'schema' can be used.");
+
+        if (request.getFormat().isPresent()) {
+          if (request.getFormat().get().requiresSchema()) {
+            checkState(
+                !request.getSchemaId().isPresent(),
+                "'schema_id=%s' cannot be used with 'serializer'.",
+                request.getSchemaId().orElse(null));
+            checkState(
+                !request.getSchemaVersion().isPresent(),
+                "'schema_version=%s' cannot be used with 'serializer'.",
+                request.getSchemaVersion().orElse(null));
```

Comment:
Except that this expression (`request.getSchemaVersion().orElse(null)` or `request.getSchemaVersion().get()`) is executed regardless of the value of the condition. So, if the condition is true, this would actually throw, even though the message would not be printed anyway.

Reply:
I would have said that we should just add a lambda-using `checkState` here, but I checked and it turns out that `checkState` is a Guava method :) Given that this is a problem just here I won't bother you with asking for a wrapper.

Reply:
Done.